### ISSUE TITLE
Accept call and can't start app when device locked or sleep ( iOS )

### ISSUE THREAD

**Issue by caglardurmus** [react-native-webrtc/react-native-callkeep#301]
# Bug report

- [x] I've checked the [example](https://github.com/react-native-webrtc/react-native-callkeep/tree/master/example) to reproduce the issue.

- Reproduced on:
 - [ ] Android
 - [x] iOS

## Description
I want to accept call and then start the application. It works perfect when the device not locked or not sleep mode but when its on locked or sleep mode, it does not start the application when i accept. I can get notification with voip. When i accept it if the user have password callkeep can't start the application.

## Expectation
When i accept the call (at sleep or locked mode) user pin his/her pass and then application start.

## Versions
    - Callkeep: 3.1.1
    - React Native: 0.62.1
    - iOS: 14
    - Phone model: iPhone 7

**Comment by nerdymind-dev** [comment]
Running into same problem. Answer goes into infinite loop since app is not available to answer/no way to launch if quit.

**Comment by caglardurmus** [comment]
Any suggestions ?

**Comment by namnm** [comment]
When phone is locked, there's no way to open app directly from PN.
https://stackoverflow.com/questions/41740139

There will be an icon in the native dialer, if user click on that icon, the phone will show unlock screen and if user unlock the phone, app will be shown. There's a config imageName when you setup CallKeep, which map to the objective C iconTemplateImageData to customize the image of that icon.

**Comment by subinsuresh** [comment]
@namnm Can you check how whatsApp handle this scenario?

**Comment by namnm** [comment]
@subinsuresh I dont use whatsapp, so I dont know its behavior. Can you add some tests and capture a video if possible? So let me sum up:
- Phone are locked with passcode (no fingerprint or face id, to make sure it is totally locked without accidentally unlock via biometric)
- Audio call (if video call, you can pass an option to the callkeep and it can show video with some UI, not sure if this is how whatsapp been doing)

**Comment by hussainsherwani** [comment]
@caglardurmus did you find any solution ? , for me it happened randomly most of the case when I open the lock app already opened and navigate to call screen but in some time it never open the application. @namnm can this be a problem #346 , if add delay what will be your suggestion ?

**Comment by caglardurmus** [comment]
> @caglardurmus did you find any solution ? , for me it happened randomly most of the case when I open the lock app already opened and navigate to call screen but in some time it never open the application. @namnm can this be a problem #346 , if add delay what will be your suggestion ?

Nope, i didn't find any solution. I used delay but its not working. It works if user answer the phone after 1.5 sec. If user answer it fast it doesn't work.

**Comment by danilvalov** [comment]
> Nope, i didn't find any solution. I used delay but its not working. It works if user answer the phone after 1.5 sec. If user answer it fast it doesn't work.

Do you have any solution for this issue?
I have the same problem: Answer/Reject buttons don't work for background/killed app state in first 1.5-5 seconds when app isn't runned yet.

**Comment by MinhThu100200** [comment]
I need a help =(( Do you have any solution for this issue?

**Comment by namnm** [comment]
@MinhThu100200 On ios, if the phone is locked, when answer the call it will display the CallKit UI.
You can put an icon from your app to the screen as below, if user click the icon, it will request for unlock and open the app afterward:
![32146144-8d545e5c-bca9-11e7-894d-b3cc516dc7ee](https://github.com/react-native-webrtc/react-native-callkeep/assets/8970908/4ae475e5-0600-44a1-8b46-4c6b431fee8e)

There could be a hacky solution: that if you enable video in RNCallKeep.displayIncomingCall
Then it will request for unlock immediately right after answer the call.

There is also another solution which I believe Whatsapp is using as the community has mentioned above.
Use PushNotification to notify incoming call instead of CallKit, then if user press the notification, it will request for unlock and open the app.
But then you could loose some of the feature of CallKit, such as notify on hold when there is incoming GSM call, or switch between multiple calls.

**Comment by MinhThu100200** [comment]
@namnm When I use IP 7 plus, everything is oke, but with others when I accept call, my app cannot open but I open my app myself. everything that I use to handle calling is running. I have no ideas about that =((((

**Comment by namnm** [comment]
@MinhThu100200 Please be more specific. When you performed those testings, was the phone locked or not, if locked with passcode or not?

Please check again and provide more details if possible, especially above questions for the iPhone7 and others you mentioned.

No one can help you if you provide little to no information at all.

**Comment by MinhThu100200** [comment]
@namnm yeah, thanks. When my app state is background or kill. And I have a phone call that is incoming. Then I answer it but my app doesn't launch. However, my handling in js code that still run. So my issue that is "accepting call but my app doesn't launch when my device still runs or locked". With iphone 7 plus, everything is okie, app always launch when I accept call. Please help me =(((

**Comment by MinhThu100200** [comment]
maybe because of version os. when I use my app with ios 16 or lower (15.x). everything is okie. But with ios 16.6, when I accept call, my app doesn't open.

**Comment by namnm** [comment]
@MinhThu100200 Please run your app in debug mode with logger and see if there's any error?
You can add this package to catch the error: https://github.com/a7ul/react-native-exception-handler
I guess your app could be crashed during the answer, adding the above package could help preventing the crash.

You can also use the Console app to view the log of your iPhone in realtime, with this you can also view the log of production bundle ipa:
https://support.apple.com/guide/console/welcome/mac
https://support.apple.com/guide/console/cnsl1012/mac

**Comment by MinhThu100200** [comment]
@namnm thank sir. I will check and follow your suggestion. thank u so much~

**Comment by Irfanwani** [comment]
> @MinhThu100200 On ios, if the phone is locked, when answer the call it will display the CallKit UI. You can put an icon from your app to the screen as below, if user click the icon, it will request for unlock and open the app afterward: ![32146144-8d545e5c-bca9-11e7-894d-b3cc516dc7ee](https://private-user-images.githubusercontent.com/8970908/267271581-4ae475e5-0600-44a1-8b46-4c6b431fee8e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxMjA1NjksIm5iZiI6MTcyNDEyMDI2OSwicGF0aCI6Ii84OTcwOTA4LzI2NzI3MTU4MS00YWU0NzVlNS0wNjAwLTQ0YTEtOGI0Ni00YzZiNDMxZmVlOGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjBUMDIxNzQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmJlZTA5ZTQ0ZjA2YzY5YjIyYTIzMWJjZTJmYTEzOGIxMWI4MWE3Yzg5MzFiZGI5N2M5ZmMwNWFhNGYxMDgwMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.n8Bkmq-T7GiFTnUUaQRONNdKpnioraiL48fzgWl5GLA)
> 
> There could be a hacky solution: that if you enable video in RNCallKeep.displayIncomingCall Then it will request for unlock immediately right after answer the call.
> 
> There is also another solution which I believe Whatsapp is using as the community has mentioned above. Use PushNotification to notify incoming call instead of CallKit, then if user press the notification, it will request for unlock and open the app. But then you could loose some of the feature of CallKit, such as notify on hold when there is incoming GSM call, or switch between multiple calls.

i think the first solution is the best way to handle this.
Let's talk about the other two solutions first (hacks infact):
1) Setting the call as video call doesn't always work as i tried it + it gives user a wrong idea about the call, so not an optimal solution.
2) Sending a push notification still requires user interaction then opening the phone (phone lock) and then the app will show.

First option is just better, because:
1) It provides native UI thus giving a better calling experience, we can simply add listeners to make all the action buttons work (or can disable them accordingly, though i don't know how, but i saw in google meet, the keypad is disabled as DTMF is not used there).
2) user doesn't need to open app to interact with the call, atleast call is setup before bringing the app to foreground, so if you have some extra features for which you want to open the app, the user can simply click the app icon shown in the native call UI, check [this](https://github.com/react-native-webrtc/react-native-callkeep/issues/320#issuecomment-2296074343) also.
3) After iOS 13, apple made it mandatory to report call to callkit when receiving a voip push, so it is a reliable solution for a calling app than using background messaging.

**Comment by gentlee** [comment]
1. If you end native call immediately after answer, make it with timeout, like 300 ms.
2. Make sure you display a video call.
